### PR TITLE
feat: create pool and create/remove liquidity position for v3

### PR DESCRIPTION
**Pull request by KeremP**
Added V3 features:

- create/get v3 pool instance
- create v3 liquidity position
- remove v3 liquidity position

### CODE REVIEW

**Comment by ErikBjare** [comment]
Very nice! Let me know when you want a review :)

Any idea why the CI chokes on the CLI tests?

**Comment by ErikBjare** [comment]
Btw, do you think this would be easy to do at the same time? https://github.com/uniswap-python/uniswap-python/issues/254

**Comment by liquid-8** [comment]
> Btw, do you think this would be easy to do at the same time? #254

Long story short:
https://stackoverflow.com/questions/71814845/how-to-calculate-uniswap-v3-pools-total-value-locked-tvl-on-chain
Due to personal reasons, I can't implement that myself at the moment. However, it seems to be not too complicated to add.

**Comment by KeremP** [comment]
> > Btw, do you think this would be easy to do at the same time? #254
> 
> Long story short: https://stackoverflow.com/questions/71814845/how-to-calculate-uniswap-v3-pools-total-value-locked-tvl-on-chain Due to personal reasons, I can't implement that myself at the moment. However, it seems to be not too complicated to add.

Sure thing, seems pretty straightforward to implement

**Comment by KeremP** [comment]
> Very nice! Let me know when you want a review :)
> 
> Any idea why the CI chokes on the CLI tests?

Took a peek at the logs for the CI, seems like provider env variable may not have set

**Comment by ErikBjare** [comment]
Ah yeah, I just now remember that secrets (like out PROVIDER credentials) aren't set on PRs.

Might get a second key, obfuscate it slightly, and use it for PRs.

**Comment by ErikBjare** [comment]
Should have fixed it in https://github.com/uniswap-python/uniswap-python/pull/266, feel free to rebase so the tests get green :)

**Comment by KeremP** [comment]
Now that we have all green :) Going to write tests for the methods that address #254 and should be good for a review

**Comment by KeremP** [comment]
@ErikBjare this is now ready for review :) 

wrt to #254 I've added a function get_tvl_in_pool() and have commented the logic (some weird hacks with the tickBitmap to get it working), but let me know if any questions.

**Comment by liquid-8** [comment]
Amazing job, that's impressive. The only thing I just can't get why
`def get_liquidity_positions(self) -> List[int]:`
instead of
`def get_liquidity_positions(address: AddressLike) -> List[int]:`

IMO if we have a scenario when this function has the reason to be implemented for our address, then we also have a similar/same scenario for others.

**Comment by KeremP** [comment]
Fair point, I agree. I'll make that change quick.

**Comment by liquid-8** [comment]
@ErikBjare just take a look at this grandeur!

File: uniswap/uniswap.py [hunk 1]
```diff
@@ -430,6 +459,7 @@ def make_trade(
                 input_token, qty, recipient, fee, slippage, fee_on_transfer
             )
         else:
+            print(input_token)
```
Comment (ErikBjare) [inline uniswap/uniswap.py:462]:
```suggestion
```

**Review by ErikBjare — APPROVED** [review]
Awesome work @KeremP! Sorry for the late review.

Found no major issues, just a couple nits. I'll fix and merge :)

File: uniswap/uniswap.py [hunk 2]
```diff
@@ -1063,6 +1093,240 @@ def remove_liquidity(self, token: str, max_token: int) -> HexBytes:
...
+        self, 
+        tokenId: int, 
+        amount0Min: int = 0, 
+        amount1Min: int = 0, 
+        deadline: int = None
+        ) -> TxReceipt:
+        """
+        remove all liquidity from the position associated w/ tokenId, collect fees, and burn token.
+        """
+        position = self.nonFungiblePositionManager.functions.positions(tokenId).call()
```
[the remark this answers is not in the document]
Comment (ErikBjare) [reply uniswap/uniswap.py:1171]: Skipping.

**Comment by ErikBjare** [comment]
Merged a squashed version in https://github.com/uniswap-python/uniswap-python/commit/9edfbaa5354b2d2df70cccb06c19a4a4c33fc64d (which was messy), now trying to address some issues during rebase.

**Comment by ErikBjare** [comment]
Alright, I can't manage to get this PR to get marked as "Merged" for some reason (keeps complaining about some conflict?).

Changes are merged though, fixing some lingering issues on master.

Thanks again for contributing @KeremP! :heart:

**Comment by KeremP** [comment]
Sounds good! Thanks for the review :)

**Comment by kethiris** [comment]
What is the difference between mint_liquidity and mint_position here? I tried to create positions using both. mint_liquidity was successful but mint_position didn't seem to create a position though the function execution was successful.

**Pull request closed without merging**